src/utils/include/utils/for_each_pair.hpp

```diff
@@ -42,15 +42,49 @@ void for_each_pair(ForwardIterator first, ForwardIterator last, BinaryOp op) {
   }
 }
 
-/*
- * @brief Range overload for for_each_pair.
- */
+/** @overload */
 template <typename ForwardRange, typename BinaryOp>
 void for_each_pair(ForwardRange &&rng, BinaryOp &&op) {
   using std::begin;
   using std::end;
   for_each_pair(begin(rng), end(rng), std::forward<BinaryOp>(op));
 }
-} // namespace Utils
 
+/**
+ * @brief Execute op for each pair of elements between (first1, last1] and
+ * (first2, last2].
+ *
+ * Diagonal elements are excluded. Pair are traversed ordered, so that
+ * for op(*it, *jt), it holds that distance(it - first) < distance(jt - first),
+ * and distance(it_n - first) < distance(it_n+1 - first) for consecutive calls.
+ */
+template <typename ForwardIterator, typename BinaryOp>
+void for_each_cartesian_pair(ForwardIterator first1, ForwardIterator last1,
+                             ForwardIterator first2, ForwardIterator last2,
+                             BinaryOp op) {
+  if (first1 == first2 and last1 == last2) {
+    for_each_pair(first1, last1, op);
+    return;
+  }
+  while (first1 != last1) {
+    for (auto it = first2; it != last2; ++it) {
+      if (*it != *first1) {
```

Comment:
I'm not sure that comparison by value is what we want here. This limits the usability to value types that are comparable without need, and the behavior is different from `for_each` (which does `it != first`, e.g. shallow comparison), which is surprising. I think in the general algorithm the pairs should not be filtered, this can be implemented in the op. One could also implement a `for_each_pair_if` on if wanted.

Reply:
I was unsure about that solution too. If we remove that conditional, then we cannot decay this function to `for_each_pair`, since that one removes diagonal elements. In fact, we probably don't want to decay in any case, because `for_each_pair` also removes the symmetric pairs.

Reply:
I've coded a `for_each_cartesian_pair_if` and removed the decay to `for_each_pair`.